packages/next/shared/lib/router/router.ts

```diff
@@ -804,29 +811,29 @@ export default class Router implements BaseRouter {
     }
 
     let forcedScroll: { x: number; y: number } | undefined
-    const { url, as, options, idx } = state
+    const { url, as, options, key } = state
     if (process.env.__NEXT_SCROLL_RESTORATION) {
       if (manualScrollRestoration) {
-        if (this._idx !== idx) {
+        if (this.key !== key) {
           // Snapshot current scroll position:
           try {
             sessionStorage.setItem(
-              '__next_scroll_' + this._idx,
+              '__next_scroll_' + this.key,
               JSON.stringify({ x: self.pageXOffset, y: self.pageYOffset })
             )
           } catch {}
 
           // Restore old scroll position:
           try {
-            const v = sessionStorage.getItem('__next_scroll_' + idx)
+            const v = sessionStorage.getItem('__next_scroll_' + key)
             forcedScroll = JSON.parse(v!)
           } catch {
             forcedScroll = { x: 0, y: 0 }
           }
         }
       }
     }
-    this._idx = idx
```

Comment:
Why was this changed from index based to a random key?

Reply:
With index, it is reset to 0 when reloading, which may cause inconsistencies.

Reply:
1. Load `/person/1` (SSR)
```
+-----------+
| /person/1 |
| idx: 0    |
+-----------+
```
2. Move to `/person/2` (pushState)
```
+-----------+-----------+
| /person/1 | /person/2 |
| idx: 0    | idx: 1    |
+-----------+-----------+
```
3. Reload `/person/2` (SSR)
```
+-----------+-----------+
| /person/1 | /person/2 |
| idx: 0    | idx: 0    |
+-----------+-----------+
```
The `idx` value and scroll position are accidentally shared between pages.